arbnode/dataposter/data_poster.go

```diff
@@ -0,0 +1,386 @@
+			"maxFeeCap", maxFeeCap,
+			"elapsed", elapsed,
+		)
+		newFeeCap = maxFeeCap
+	}
+
+	return newFeeCap, newTipCap, nil
+}
+
+func (p *DataPoster[Meta]) PostTransaction(ctx context.Context, dataCreatedAt time.Time, nonce uint64, meta Meta, to common.Address, calldata []byte, gasLimit uint64) error {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	feeCap, tipCap, err := p.getFeeAndTipCaps(ctx, nil, dataCreatedAt)
+	if err != nil {
+		return err
+	}
+	inner := types.DynamicFeeTx{
+		Nonce:     nonce,
+		GasTipCap: tipCap,
+		GasFeeCap: feeCap,
+		Gas:       gasLimit,
+		To:        &to,
+		Value:     new(big.Int),
+		Data:      calldata,
+	}
+	fullTx, err := p.auth.Signer(p.auth.From, types.NewTx(&inner))
+	if err != nil {
+		return err
+	}
+	queuedTx := queuedTransaction[Meta]{
+		Data:            inner,
+		FullTx:          fullTx,
+		Meta:            meta,
+		Sent:            false,
+		Created:         dataCreatedAt,
+		NextReplacement: time.Now().Add(p.replacementTimes[0]),
+	}
+	fmt.Printf("sending nonce: %v\n", inner.Nonce)
+	return p.sendTx(ctx, nil, &queuedTx)
+}
+
+// the mutex must be held by the caller
+func (p *DataPoster[Meta]) saveTx(ctx context.Context, prevTx *queuedTransaction[Meta], newTx *queuedTransaction[Meta]) error {
+	if prevTx != nil && prevTx.Data.Nonce != newTx.Data.Nonce {
+		return fmt.Errorf("prevTx nonce %v doesn't match newTx nonce %v", prevTx.Data.Nonce, newTx.Data.Nonce)
+	}
+	return p.queue.Put(ctx, newTx.Data.Nonce, prevTx, newTx)
+}
+
+func (p *DataPoster[Meta]) sendTx(ctx context.Context, prevTx *queuedTransaction[Meta], newTx *queuedTransaction[Meta]) error {
+	err := p.saveTx(ctx, prevTx, newTx)
+	if err != nil {
+		return err
+	}
+	err = p.client.SendTransaction(ctx, newTx.FullTx)
+	if err == nil {
+		log.Info("DataPoster sent transaction", "nonce", newTx.FullTx.Nonce(), "hash", newTx.FullTx.Hash(), "feeCap", newTx.FullTx.GasFeeCap())
+		newerTx := *newTx
+		newerTx.Sent = true
+		err = p.saveTx(ctx, newTx, &newerTx)
+	} else {
+		log.Warn("DataPoster failed to send transaction", "err", err, "nonce", newTx.FullTx.Nonce(), "feeCap", newTx.FullTx.GasFeeCap())
+	}
+	return err
+}
+
+// the mutex must be held by the caller
+func (p *DataPoster[Meta]) replaceTx(ctx context.Context, tx *queuedTransaction[Meta]) error {
+	newFeeCap, newTipCap, err := p.getFeeAndTipCaps(ctx, tx.Data.GasTipCap, tx.Created)
+	if err != nil {
+		return err
+	}
+
+	desiredFeeCap := newFeeCap
+	maxFeeCap := new(big.Int).Div(p.balance, new(big.Int).SetUint64(tx.Data.Gas))
+	newFeeCap = arbmath.BigMin(newFeeCap, maxFeeCap)
+	minNewFeeCap := arbmath.BigMulByBips(tx.Data.GasFeeCap, minRbfIncrease)
+	newTx := *tx
+	if newFeeCap.Cmp(minNewFeeCap) < 0 {
+		if desiredFeeCap.Cmp(minNewFeeCap) >= 0 {
+			log.Error(
+				"lack of L1 balance prevents posting transaction with a higher fee cap",
+				"balance", p.balance,
+				"gasLimit", tx.Data.Gas,
+				"desiredFeeCap", desiredFeeCap,
+				"maxFeeCap", maxFeeCap,
+			)
+		}
```

Comment:
Hmm, yeah, my assumption was that if sendRawTransaction returned no error it'd been broadcast, but you're right that it could easily get lost. I'll rebroadcast here.